src/nvim/statusline.c

```diff
@@ -332,31 +334,54 @@ static void win_redr_custom(win_T *wp, bool draw_winbar, bool draw_ruler)
   int len = (int)strlen(buf);
   int start_col = col;
 
-  // Draw each snippet with the specified highlighting.
-  screengrid_line_start(grid, row, 0);
+  if (!external) {
+    // Draw each snippet with the specified highlighting.
+    screengrid_line_start(grid, row, 0);
+  }
 
-  int curattr = attr;
   char *p = buf;
-  for (int n = 0; hltab[n].start != NULL; n++) {
-    int textlen = (int)(hltab[n].start - p);
-    // Make all characters printable.
-    size_t tsize = transstr_buf(p, textlen, transbuf, sizeof transbuf, true);
-    col += grid_line_puts(col, transbuf, (int)tsize, curattr);
-    p = hltab[n].start;
-
-    if (hltab[n].userhl == 0) {
+  int curgroup = 0;
+  int curattr = attr;
+  Array content = ARRAY_DICT_INIT;
+  for (stl_hlrec_t *sp = hltab;; sp++) {
+    int textlen = (int)(sp->start ? sp->start - p : buf + len - p);
+    // Make all characters printable. Use an empty string instead of p, if p is beyond buf + len.
+    size_t tsize = transstr_buf(p >= buf + len ? "" : p, textlen, transbuf, sizeof transbuf, true);
+    if (!external) {
+      col += grid_line_puts(col, transbuf, (int)tsize, curattr);
+    } else {
+      Array chunk = ARRAY_DICT_INIT;
+      ADD(chunk, INTEGER_OBJ(curattr));
+      ADD(chunk, STRING_OBJ(cbuf_as_string(xmemdupz(transbuf, tsize), tsize)));
+      ADD(chunk, INTEGER_OBJ(curgroup));
```

Comment:
Caveat here that this does not convey combined highlight groups. Would need to adapt the "message chunks" API signature to include that.

Just making it `[attr_id, text, group_id, combine_id]` would technically be non-breaking I suppose. IDK if we would then include a 0 `combine_id` in other places. And/or if a breaking change removing `attr_id` and making `group_id` an array (so it can be passed to `nvim_buf_set_extmark()` directly) should be considered @bfredl.